packages/server/src/integrations/base/clientCache.ts

```diff
@@ -0,0 +1,65 @@
+import { Duration } from "@budibase/backend-core"
+import LRUCache from "lru-cache"
+
+interface Merkle {
+  hash: (json: Record<string, any>) => string
+  stringify: (json: Record<string, any>) => string
+}
+
+interface MerkleConstructor {
+  new (): Merkle
+}
+
+// no typescript definition - fill it in
+const { MerkleJson } = require("merkle-json") as {
+  MerkleJson: MerkleConstructor
+}
+
+const jsonHash = new MerkleJson()
+
+type ClientCacheValue<T> = {
+  destroy: (client: T) => Promise<void>
+  // can't really know the type of the client, they are all so different
+  client: T
+}
+
+const clientCache = new LRUCache<string, ClientCacheValue<any>>({
+  // max 50 active clients per service
+  max: 50,
+  ttl: Duration.fromMinutes(10).toMs(),
+  updateAgeOnGet: true,
+  dispose: async (value: ClientCacheValue<any>) => {
+    await value.destroy(value.client)
+  },
+})
+
+// this class is used to define the implementation of disposal and type for each DS+
+export class ClientCache<T> {
+  destroy: (client: T) => Promise<void>
+
+  constructor(destroy: (client: T) => Promise<void>) {
+    this.destroy = destroy
+  }
+
+  hash(config: Record<string, any>) {
+    return jsonHash.hash(config)
+  }
+
+  has(configHash: string): boolean {
+    return clientCache.has(configHash)
+  }
```

Comment:
We could remove `has` entirely and rely on `get` returning undefined when the item is not found. This removes a check from every time we try to get a connection.

I think this also solves a potential problem where `has` could return true, then the TTL could expire, then `get` returns `undefined`.

Reply:
Removed has check.